crypto/rsa/rsa_sign.c

```diff
@@ -16,60 +16,189 @@
+    MD_CASE(sha512_256);
+    MD_CASE(sha3_224);
+    MD_CASE(sha3_256);
+    MD_CASE(sha3_384);
+    MD_CASE(sha3_512);
+    default:
+        return NULL;
+    }
+}
+
+/*
+ * Alternative DigestInfo encoding without parameters.
+ * The parameters may be not present (i.e. not NULL) when a SHA digest is used.
+ */
+static const unsigned char *digestinfo_encoding_nn(int nid, size_t *len)
+{
+    switch (nid) {
+#ifndef FIPS_MODE
+    MD_CASE_NN(sha1);
+#endif
+    MD_CASE_NN(sha224);
+    MD_CASE_NN(sha256);
+    MD_CASE_NN(sha384);
+    MD_CASE_NN(sha512);
+    MD_CASE_NN(sha512_224);
+    MD_CASE_NN(sha512_256);
+    MD_CASE_NN(sha3_224);
+    MD_CASE_NN(sha3_256);
+    MD_CASE_NN(sha3_384);
+    MD_CASE_NN(sha3_512);
+    default:
+        return NULL;
+    }
+}
+
 /* Size of an SSL signature: MD5+SHA1 */
 #define SSL_SIG_LENGTH  36
 
 /*
  * encode_pkcs1 encodes a DigestInfo prefix of hash |type| and digest |m|, as
  * described in EMSA-PKCS1-v1_5-ENCODE, RFC 3447 section 9.2 step 2. This
  * encodes the DigestInfo (T and tLen) but does not add the padding.
+ * 
+ * When verifying, decrypt_len indicates the size of the decrypted DigestInfo
+ * and is used to determine if the alternative encoding should be used for
+ * comparison.  When signing, decrypt_len is zero.
  *
  * On success, it returns one and sets |*out| to a newly allocated buffer
  * containing the result and |*out_len| to its length. The caller must free
  * |*out| with |OPENSSL_free|. Otherwise, it returns zero.
  */
-static int encode_pkcs1(unsigned char **out, int *out_len, int type,
-                        const unsigned char *m, unsigned int m_len)
+static int encode_pkcs1(unsigned char **out, size_t *out_len,
+                        size_t decrypt_len, int type,
+                        const unsigned char *m, size_t m_len)
 {
-    X509_SIG sig;
-    X509_ALGOR algor;
-    ASN1_TYPE parameter;
-    ASN1_OCTET_STRING digest;
-    uint8_t *der = NULL;
-    int len;
-
-    sig.algor = &algor;
-    sig.algor->algorithm = OBJ_nid2obj(type);
-    if (sig.algor->algorithm == NULL) {
+    size_t di_prefix_len, dig_info_len;
+    const unsigned char *di_prefix;
+    unsigned char *dig_info;
+
+    if (type == NID_undef) {
         RSAerr(RSA_F_ENCODE_PKCS1, RSA_R_UNKNOWN_ALGORITHM_TYPE);
         return 0;
     }
-    if (OBJ_length(sig.algor->algorithm) == 0) {
+    di_prefix = digestinfo_encoding(type, &di_prefix_len);
+    if (di_prefix == NULL) {
         RSAerr(RSA_F_ENCODE_PKCS1,
                RSA_R_THE_ASN1_OBJECT_IDENTIFIER_IS_NOT_KNOWN_FOR_THIS_MD);
         return 0;
     }
-    parameter.type = V_ASN1_NULL;
-    parameter.value.ptr = NULL;
-    sig.algor->parameter = &parameter;
-
-    sig.digest = &digest;
-    sig.digest->data = (unsigned char *)m;
-    sig.digest->length = m_len;
+    dig_info_len = di_prefix_len + m_len;
+    if (decrypt_len != 0 && dig_info_len != decrypt_len) {
+        /* Use alternate encoding (parameters not present) */
```

Comment:
How was this case handled in the previous code? I don't see it? Or is this a new feature?

Reply:
Well, probably more like an old feature. The support for alternate encodings with the optional parameters field omitted was based on `fips_rsa_sign.c` from the OpenSSL FIPS Object Module, with a Copyright year of 2007. After a more careful reading of that RFC and the more recent
[RFC 8017](https://tools.ietf.org/html/rfc8017) I now think support for the alternative encodings is a mistake and will remove them.

The thinking behind the alternate encodings was probably based on the following paragraphs in [Appendix B.1 of RFC 3447](https://tools.ietf.org/html/rfc3447#appendix-B.1).

> The parameters field associated with id-md2 and id-md5 in a value of
> type AlgorithmIdentifier shall have a value of type NULL.
> 
> The parameters field associated with id-sha1, id-sha256, id-sha384,
> and id-sha512 should be omitted, but if present, shall have a value
> of type NULL.

... however that doesn't appear to apply to the AlgorithmIdentifier within a DigestInfo.
[RFC 8017 Appendix A.2.4. RSASSA-PKCS-v1_5](https://tools.ietf.org/html/rfc8017#appendix-A.2.4) is clear that "For each OID, the parameters field associated with this OID in a value of type AlgorithmIdentifier SHALL have a value of type NULL."

[Appendix B.1 of RFC 8017](https://tools.ietf.org/html/rfc8017#appendix-B.1) also has the
following paragraphs:

> This is to align with the definitions originally promulgated by NIST.
> For the SHA algorithms, implementations MUST accept
> AlgorithmIdentifier values both without parameters and with NULL
> parameters.
> 
> Exception: When formatting the DigestInfoValue in EMSA-PKCS1-v1_5
> (see Section 9.2), the parameters field associated with id-sha1,
> id-sha224, id-sha256, id-sha384, id-sha512, id-sha512/224, and
> id-sha512/256 shall have a value of type NULL.  This is to maintain
> compatibility with existing implementations and with the numeric
> information values already published for EMSA-PKCS1-v1_5, which are
> also reflected in IEEE 1363a [IEEE1363A].

... where EMSA-PKCS1-v1_5 is the encoding method upon which the Signature Scheme
(RSASSA-PKCS-v1_5) is based.

Reply:
Regardless of if it is actually correct or not, I think it should still handle both cases for the decode - just in case there is a backwards compatibility issue.

Reply:
The existing OpenSSL code (apart from the FOM) doesn't handle both cases, so I guess there is no known compatibility issue. The old FOM doesn't cause a backwards compatibility issue since it was still using the NULL parameters case when signing.